Fix incorrect course name displayed for CCX courses.

taranjeet opened the pull request:
This commit updates the course name for CCX course at Info, Course
and Instructor tabs.

**JIRA tickets**: [OSPR-2557](https://openedx.atlassian.net/browse/OSPR-2557)

**Screenshots**: Attached are some before/after screenshots to show the change

*Note* : The text in the **green box** is the **correct** one whereas the text in the **red box** is the **incorrect** one.

**Course Tab (before)**

<img width="752" alt="Course Tab Before" src="https://user-images.githubusercontent.com/4302268/44440455-c4a45d00-a5e5-11e8-98f4-3c4938577d2d.png">

**Course Tab (after)**

<img width="625" alt="Course Tab After" src="https://user-images.githubusercontent.com/4302268/44440498-f1587480-a5e5-11e8-8b77-3d6294562f78.png">

**Instructor Tab (before)**

<img width="679" alt="Instructor Tab Before" src="https://user-images.githubusercontent.com/4302268/44440551-2b297b00-a5e6-11e8-8193-7cf15e8e47cc.png">

**Instructor Tab (after)**

<img width="716" alt="Instructor Tab After" src="https://user-images.githubusercontent.com/4302268/44440559-35e41000-a5e6-11e8-9ae1-92fb64ed7bf9.png">

**Sandbox URL**: 

LMS: https://pr18808.sandbox.opencraft.hosting/
Studio: https://studio-pr18808.sandbox.opencraft.hosting/

**Testing instructions**:

1. Go to Course Tab ([here](https://pr18808.sandbox.opencraft.hosting/courses/ccx-v1:UniversityOfOpenEducation+UEOEC01+2018_Q4+ccx@1/course/)) and verify that the heading(h2) with class of "hd hd-3 page-title" contains the name of child course
2. Go to Instructor Tab ([here](https://pr18808.sandbox.opencraft.hosting/courses/ccx-v1:UniversityOfOpenEducation+UEOEC01+2018_Q4+ccx@1/instructor#view-course_info)) and verify that the value corresponding to "Course Name" heading in "Basic Course Information" is the name of child course and not parent course.

**Reviewers**
- [ ] @itsjeyd 
- [ ] @ormsbee 

**Settings**
```yaml
EDXAPP_FEATURES:
  CUSTOM_COURSES_EDX: true
```

edx-webhook commented:
Thanks for the pull request, @taranjeet! I've created [OSPR-2557](https://openedx.atlassian.net/browse/OSPR-2557) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

mduboseedx commented:
@taranjeet Thank you for the contribution! @ormsbee This is a bug fix for CCX courses. Could you give this a quick look over when you have a chance?

itsjeyd commented on lms/templates/courseware/info.html:
```diff
@@ -65,7 +66,14 @@ <h2 class="hd hd-2 page-title">
               % if show_subtitle:
-                <div class="page-subtitle">${course_subtitle}</div>
+                <%
+                  display_course_subtitle = course_subtitle
```
@taranjeet There's no need to introduce another variable here, I think. You can just remove this line and do

```python
course_subtitle = ccx.display_name
```

below.

taranjeet replied:
@itsjeyd : This is fixed. Thanks for pointing it out!

taranjeet replied:
@itsjeyd : I think we have to live with this only. From the [docs](http://docs.makotemplates.org/en/latest/runtime.html#context-variables), we can see that this requires changing the way context is passed into render. 

> Another facet of the Context is that its dictionary of variables is immutable

PS: Read the whole part after this.upto heading "Context Methods and Accessors".

If there is any suggestion or I am missing something, please let me know.

Thanks :)

itsjeyd replied:
@taranjeet Ah, OK! That's a subtlety I wasn't aware of. Thanks for clarifying, there's no need to change this, then.

ormsbee commented:
Sorry, I think I might be misunderstanding. These are my assumptions:

1. CCX stores overrides for arbitrary (course, location, field) tuples in the database. It's the basis of all content overrides for CCX, and operates at the modulestore level.
2. We can use that same mechanism to override the `display_name` attribute for the location of the root CourseModule.

The first pass at this PR did the explicit check/override at the view/template layer.

The second pass creates a new method in CourseModule to do the "if CCX" check and then conditionally pull the course name from CCX's course information. This does help insulate more systems from having to understand CCX, but would force the CourseModule to understand it.

The issue with this approach is that there are other things in the system that still call `.display_name` or will otherwise pull that field down via iteration over .fields(). In which case we'll continue to get mismatches where a CCX course shows the right name in these two places, but the root course name in N other places.

What I'm requesting is that you push the work one more layer down, so that it's stored as a field override of the root course block, just like you would store overrides for other content that CCX coaches can tweak. This may mean that there will be some data duplication and that we'll store CCX names in both the table they're in now as well as a second copy in the field overrides table. Possibly even a migration script. But it would keep those CCX content modifications isolated to the modulestore layer and not let it bubble up to other systems and cause difficult-to-debug name-sync errors. This might seem like more trouble than it's worth, but we've spent years slowly bottling up the complexity of maintaining CCX-specific logic -- the more it leaks out, even in small ways, the larger the overall maintenance cost of CCX becomes.

Now, given it's the modulestore, I wouldn't be surprised if there were some really horrible reason why this approach won't actually work. But if so, I'd really like to understand it better.

Thank you.

taranjeet commented:
@ormsbee : Thanks a lot for your input. I will look into the approach and update you accordingly. :)

taranjeet commented:
@ormsbee : I have now implemented the functionality to store display name as an override field. This will make sure that any child course saved in the future will have the correct display name. But for existing courses, we need to add entries in the database, for which a migration script needs to be written and run on production servers. 
Please let me know your inputs/thoughts etc.

Thanks

ormsbee commented:
@taranjeet: Please create a data migration that scans through existing CCX courses and sets the `display_name` overrides. Makes sure that your data migration also invokes `SignalHandler.course_publish.send()` with each CCX course key so that things like CourseOverviews are properly updated.

taranjeet commented:
@ormsbee Thank you :) I will update you about the approach.

ormsbee commented:
@taranjeet: FWIW, if you want to split the code that makes it work for new courses and the code that does the data migration into two PRs, that's an option. But I would ask that you add a test that verifies this override actually takes place and works after creating a new CCX.

taranjeet commented:
@ormsbee : Thanks for the suggestion. I will let you know about how I am planning to approach.

CC: @itsjeyd

taranjeet commented:
@itsjeyd @ormsbee 

I have added tests to check if the child course name is getting correctly saved or not. I have also added a data migration, which will populate the value of `display_name` for previous course. However there are some notes

* Migrations are reversible in the sense that only the entry from `django_migrations` table gets deleted, but the corresponding rows in `ccx_ccxfieldoverride` are not removed.

* When writing a data migration, we need to use `apps.get_model` to get model. But when I tried using `override_field_for_ccx` there is this function named `_get_overrides_for_ccx` which uses `CcxFieldOverride` being imported from the app(`ccx.models`). I tried the patching the model load function using this [answer](https://stackoverflow.com/a/25981899/2534102), but it didn't worked hence I had to copy all the functions in the data migration and pass an additional argument of `CcxFieldOverride` which is loaded via `apps.get_model`. If there is any workaround for this, please let me know. :)

* I manually checked the migration by first apply, then unapply and then applying again. No rows were added twice.

* `_clean_ccx_key` uses `CCXBlockUsageLocator`, which I skipped passing (first loading via `apps.get_model`) because the if check was not getting true.

Thanks!

itsjeyd commented:
@taranjeet 

> I have added tests to check if the child course name is getting correctly saved or not.

The test code that you added looks good to me.

> Migrations are reversible in the sense that only the entry from `django_migrations` table gets deleted, but the corresponding rows in `ccx_ccxfieldoverride` are not removed.

Keeping the overrides for `display_name` when migrating backwards sounds fine to me, but I'll let @ormsbee provide final confirmation here.

> ... hence I had to copy all the functions in the data migration and pass an additional argument of `CcxFieldOverride` which is loaded via `apps.get_model`.

It would be good to find a way that doesn't involve duplicating all of those functions in the file that defines the data migration. If importing `override_field_for_ccx` is not an option, one possible approach would be to identify the parts of the code that are relevant for the changes that the data migration needs to make, and only recreate those in the file that defines the data migration.

More specifically: We're only interested in setting `display_name` for existing courses here. The functions that you copied over are generic and support creating overrides for other types of fields. We don't need all of that logic here. We just need the bits and pieces that allow us to create a proper override for `display_name`. With that in mind, once you analyze the functions in more detail you might even find that it's not necessary to call `get_override_for_ccx` at all.

CC @ormsbee

ormsbee commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -394,6 +394,10 @@ def test_create_ccx(self, ccx_name='New CCX'):
 
+        # check if the course display name is properly set
+        course_display_name = get_override_for_ccx(ccx, self.course, 'display_name')
+        self.assertEqual(course_display_name, ccx_name)
```
This is fine to test, but the interface we're really interested in is the modulestore. Please check that when you ask for CCX Course's root node from the modulestore and ask for its display name, it gives you the overridden name and not the name of the root course.

taranjeet replied:
Hi @ormsbee  I checked for a particular course on my local devstack and found out
that `display_name` is not getting stored in the modulestore. Infact none of the
fields(start, due, max_student_enrollments_allowed) is getting saved/updated.
Here is the code snippet which I used to verify.

```python
>>> from xmodule.modulestore.django import modulestore
>>> store = modulestore()

>>> ccx = CustomCourseForEdX.objects.get()
>>> ccx.display_name
Economics Introduction Child Course
>>> ccxm = store.get_course(ccx.course_id)
>>> ccxm.display_name
Introduction to Economics Parent Course

>>> ccx.start
datetime.datetime(2018, 8, 27, 6, 22, 43, tzinfo=tzutc())
>>> ccxm.start
datetime.datetime(2030, 1, 1, 0, 0, tzinfo=<bson.tz_util.FixedOffset object at 0x7f9bc42ae9d0>)
```

We can see that `ccx` has the right start date, where as course object from `modulestore` has the
default date.

I did some more findings and figured that `grading_policy` is getting stored via `override_field_for_ccx`.
I tried updating it, but could not see the reflected changes

```python
# assume that `new_grading_policy` variables holds a proper json object but different from `ccxm.grading_policy`
>>> from ccx_keys.locator import CCXLocator
>>> from xmodule.modulestore.django import SignalHandler
>>> from lms.djangoapps.ccx.overrides import override_field_for_ccx
>>> from courseware.courses import get_course_by_id

>>> course = get_course_by_id(ccx.course_id, depth=None)
>>> override_field_for_ccx(ccx, course, 'grading_policy', new_grading_policy)
>>> ccxm.grading_policy
# still the same object

# even after re-fetching the object from `modulestore`
>>> ccxm = store.get_course(ccx.course_id)
>>> ccxm.grading_policy
```

If I write the test, it will fail because the modulestore wont give the updated name.

CC @itsjeyd

ormsbee replied:
The `course_id` attribute of the `CustomCourseForEdX` model is the `course_id` of the root course. But if you do:

```python
ccx = CustomCourseForEdX.objects.get()
ccxm = store.get_course(ccx.locator)
```

Then `ccxm` should be pulling the CCX course with overrides from the modulestore and should hopefully have the correct `display_name`.

taranjeet replied:
@ormsbee : I have added [test](https://github.com/edx/edx-platform/pull/18808/files#diff-1f353962b6fa14872960bdbb72548764R401) for modulestore, but it is still giving the [old course name](https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/56133/testReport/junit/lms.djangoapps.ccx.tests.test_views/TestCoachDashboard/test_create_ccx/)

ormsbee commented on lms/djangoapps/ccx/migrations/0006_set_display_name_as_override.py:
```diff
@@ -0,0 +1,155 @@
+            course_key=CCXLocator.from_course_locator(course.id, unicode(ccx.id))
+        )
+        for rec, response in responses:
+            log.info('Signal fired when course is published. Receiver: %s. Response: %s', rec, response)
```
Please specify the course that's being published.

taranjeet replied:
This is fixed.

ormsbee commented:
> Keeping the overrides for display_name when migrating backwards sounds fine to me, but I'll let @ormsbee provide final confirmation here.

Yeah, that's fine. Preferable, even.

> It would be good to find a way that doesn't involve duplicating all of those functions in the file that defines the data migration. If importing override_field_for_ccx is not an option, one possible approach would be to identify the parts of the code that are relevant for the changes that the data migration needs to make, and only recreate those in the file that defines the data migration.

I agree that it would be nice to reuse existing functionality or to copy just the parts we need (remembering also that we need to handle the "the display name is already there" scenario). That being said, if it's a pain to do so, I won't consider it a blocker to merging. It's a bit ugly to have duplication like that, but it's well tested code and it's unlikely that we'll be digging through it again after the migration runs.

taranjeet commented:
@itsjeyd @ormsbee Thanks for your inputs. Let me see what can be done and I will update you accordingly. :)

taranjeet commented:
@ormsbee @itsjeyd : I have fixed the issue regarding code duplication part. It now contains only the essential and required piece of code. 
However, I am facing some issue with the test already mentioned [here](https://github.com/edx/edx-platform/pull/18808#discussion_r215510113). Regarding the failing test case, I thought that may be the signal is not getting published and hence I added `ENABLED_SIGNALS = ['course_published']` but still no luck. I also tried looking for ccx and ccx instance from modulestore in [test_ccx_modulestore](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/ccx/tests/test_ccx_modulestore.py) but again no luck over there as well. Any help for what am I missing?

Thanks :)

itsjeyd commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -393,6 +393,13 @@ def test_create_ccx(self, ccx_name='New CCX'):
+
+        # check if the course display name is properly set in modulestore
+        course_display_name = self.mstore.get_course(ccx.locator).display_name
+        self.assertEqual(course_display_name, ccx_name)
```
@taranjeet From a high level I'm not sure why this check would be failing. Did you test these steps manually (by doing what you did [earlier](https://github.com/edx/edx-platform/pull/18808#discussion_r215274883), replacing relevant parts with [this code](https://github.com/edx/edx-platform/pull/18808#discussion_r215274883)), to see if that would result in the expected behavior?

CC @ormsbee

ormsbee replied:
@taranjeet: Yeah, it's not clear to me how that test is failing if the feature itself is working. If you've gone step by step through it in a debugger, can you describe where it's going wrong? Not finding the override? Not using the override? Grabbing a cached version from somewhere in some way?

@itsjeyd: I think I'm good with the PR once the test passes. Please let me know if you need me for more review. Otherwise, I'll wait until the test passes and you give a thumbs up, and then merge.

taranjeet replied:
@ormsbee @itsjeyd : I have figured out what was missing. We need to override some settings so that the modulestore can work properly. The code is updated now.

taranjeet commented:
@itsjeyd : Thanks a lot for the detailed review. I will be addressing it and then I will let you know.
Regarding the test case, I will be doing a detailed analysis of why it is failing and then fixing it.
@ormsbee : Thanks for the review. I will dig why it is failing and update accordingly.

taranjeet commented:
@itsjeyd : I have addressed all your review concerns. Please review the PR. Thank you. :)

taranjeet commented:
@itsjeyd I have squashed the commit into one and updated the sandbox. Please verify if the changes are working fine or not. (from my side, I created this [course](https://pr18808.sandbox.opencraft.hosting/courses/ccx-v1:OpenUniversity+GEO01+2018_Q4+ccx@2/course/) just to be sure that changes are working fine or not: it is working fine )

edx-webhook commented:
@taranjeet 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18808&entry.1484655318=https://github.com/taranjeet&entry.752974735=2018-08-20+09:29&entry.1917517419=2018-09-13+14:53&entry.2133058324=Yes) so we can improve your experience in the future.

ormsbee commented:
Thanks folks! :-)

ormsbee commented:
@taranjeet, @itsjeyd: I'll look into the e2d failures.

taranjeet commented:
@itsjeyd @ormsbee : Thanks for the review and merge. :)

itsjeyd commented:
@ormsbee Thanks, let us know if you need anything re: e2e failures.

ormsbee commented:
@taranjeet, @itsjeyd: Hi folks. When running the data migration on one of our instances, we ran into this error:

```
        "      File \"/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/courses.py\", line 77, in get_course_by_id", 
        "        raise Http404(\"Course not found: {}.\".format(unicode(course_key)))", 
        "    Http404: Course not found: course-v1:DavidsonNext+Ex_Calc+1T2015."
```

This can happen when a CCX course is created on top of a root course that is later deleted (so CCX exists, but root course does not). Now, this is clearly not a common case, and we're triaging it via different means. But for the sake of anyone who might run this migration in Ironwood, could you please create a new PR that adds a try/catch block to account for this scenario and log error messages about it (but not fail)?

itsjeyd commented:
@ormsbee Thanks for the report! Sure, we'll follow up with another PR.

taranjeet commented:
@itsjeyd @ormsbee : I have opened [PR 18964](https://github.com/edx/edx-platform/pull/18964). Please review. :)

ormsbee commented:
Hi folks. Thank you again for #18964 -- it's a big help. This migration ran fine on our edx.org environment, but we were unable to run it on another environment where we have 1300+ CCX courses and had to fake the migration there in order to unblock the release. I'm going to post the findings of that here shortly because I think it's useful to know for migrations and CCX going forward, and because I don't want that information trapped somewhere deep in the bowels of our wiki where nobody will ever find it. Please don't take this as a criticism of your work here. I completely missed these things in the review process, and I think some of them are quite unintuitive, even for folks who have been working on CCX for a while.j

ormsbee commented:
# Issues Migrating Large Numbers of CCX Courses

We ran into various issues when running this migration on an environment with over 1300 CCX courses.

## Deleted Underlying Courses

It did not account for the failure scenario where an entry exists for a CCX course but the root course that the CCX extends has been deleted. Course deletion is just not very well supported in general in edx-platform. It's not clear what the correct long term behavior should be for this scenario, but in the short term this error case is now handled for the purposes of this data migration by #18964.

We did a manual deletion to get rid of these CCX entries (they were early test courses anyhow). This was complicated by foreign key constraints to the `ccx_ccxmembership` table, which also needed to be manually removed. This table was from an early incarnation of CCX where it stored its enrollments in a separate table instead of using the standard one that other courses use. Those entries were not deleted in the original data migration of CCX enrollments because we wanted the ability to rollback quickly. The entire table should have been dropped in a subsequent migration, but it looks like somehow those migrations are missing or were possibly collapsed at some point. In any event, it is not created on fresh installs but remains in our production databases.

## Logging Inconsistency

The logging for this is a little confusing:

```python
        # Publish change
        responses = SignalHandler.course_published.send(
            sender=ccx,
            course_key=CCXLocator.from_course_locator(course.id, unicode(ccx.id))
        )
        for rec, response in responses:
            log.info(
                'Signal fired when course is published. Course %s. Receiver: %s. Response: %s',
                ccx.course_id,
                rec,
                response
            )
```

Now this code does the right thing in sending course publishes for CCX courses. But the log message outputs `ccx.course_id` which is the `course_id` of the root course, not the CCX. This caused a bit of confusion when trying to search the logs and understand what happened.

The funny thing is that sending a publish signal for the root course should actually work because CCX echoes out all publishes of root courses to their respective child CCX courses. It would just be really slow as there would be many redundant publishes.

## Transaction Wrapping

Data migrations are implicitly wrapped in a big transaction, which makes sense since you'd normally want to rollback entirely if there is an error. But this is problematic for the "commit this data change and then trigger publish" pattern we use here because those publish signals are meant to be picked up by worker processes that are not in the same transaction and therefore won't see the new data.

This migration still appears to work on smaller installs because the most important listener for this is Course Overviews, which runs in-proc instead of listening asynchronously. I suspect this is actually wrong on the part of Course Overviews, but it's definitely behavior that could change at some point.

**But it gets even crazier.** Because we've run into transaction related issues around Course Overviews before, that code actually very carefully wraps things in its own transactions for commit and error handling. So we're now in a state where many Course Overviews actually have been updated to reflect the CCX overridden `display_name` for the course, but the CCX field data overrides themselves were rolled back leaving no trace of that. Unfortunately, as soon as actual publishes occur for these courses (e.g. someone edits them), Course Overviews will be updated with the wrong value because the field override for the Course's `display_name` will be missing.

Longer term action items:

* Make this migration [non-atomic](https://docs.djangoproject.com/en/1.11/howto/writing-migrations/#non-atomic-migrations).
* Possibly change it so that we just always write the `display_name` override on a CCX course whenever we do a save/publish.

## Memory Leak

The process got a little more than halfway through the courses before it ran out of memory (~8 GB) and was killed by the OS. This likely means we've got another ugly memory leak somewhere in modulestore. This is a deeper issue that we need to tackle and is likely coming from somewhere deep in the modulestore (and some investigation is being done now). If the transaction wrapping were more granular, it's possible we could restart this data migration a few times and eventually it would succeed.

ormsbee commented:
Actually, having slept on this, I realize that the easiest way around the transaction issue is to just make sure we do all our database writes of the overrides in a transaction, and only after we've committed them all do we cycle through again and send the publish signals. That way, even if the publish processing fails, all the overrides are committed and the name changes will eventually work their way out as these courses are worked on and individually published.

itsjeyd commented:
@ormsbee Thanks a lot for the detailed report on how the migration went on your end. And sorry that it ended up generating some unexpected overhead! We'll make sure to allocate time for reviewing any follow-up PRs that you send.

bradenmacdonald commented:
Wow, what a tangled web. Sorry you had to deal with this, @ormsbee - and thanks for the writeup!

itsjeyd commented:
Related PRs (for future reference):

* #19022: Remove course publish from CCX data migration.
* #18020: Add simulate_publish management command.
